\documentclass[a4paper,12pt]{article}

\usepackage{graphicx}
\usepackage[small]{caption}
\usepackage[utf8]{inputenc}
\usepackage[T1]{fontenc}
\usepackage{setspace}
\usepackage[outdir=./]{epstopdf}
\usepackage[a4paper, top = 2cm, bottom = 2cm]{geometry}
\usepackage[table,xcdraw]{xcolor}
\usepackage{hyperref}
\hypersetup{colorlinks=true,linkcolor=blue,urlcolor=blue}

\title{Supplementary Figures.}


\setlength{\oddsidemargin}{-0cm}
\setlength{\textwidth}{16cm}



\renewcommand{\figurename}{\textbf{Fig.}}
\renewcommand{\thefigure}{\textbf{S\arabic{figure}}}
\renewcommand{\thetable}{\textbf{S\arabic{table}}}

\begin{document}

\onehalfspacing

\begin{center}
	\Large
	\noindent \textbf{Manuscript title}

	\normalsize

	\vskip 3cm

\end{center}

%\noindent Eric R. Lucas$^{1}$* 

\noindent Author M. Surname \ldots and The \textit{Anopheles gambiae} 1000 Genomes Consortium
 
\vskip 2cm 


\section*{Electronic Supplementary Material \\ Supplementary figures and tables}

\clearpage

\begin{figure}[h]
	\begin{center}
		\makebox{\includegraphics*{../../CNV_analysis/Ace1_diagnostic_read_CNVs.png}}
		\vskip 0.4cm
		\makebox{\includegraphics*{../../CNV_analysis/Cyp6aap_diagnostic_read_CNVs.png}}
		\vskip 0.4cm
		\makebox{\includegraphics*{../../CNV_analysis/Cyp6mz_diagnostic_read_CNVs.png}}
		\vskip 0.4cm
		\makebox{\includegraphics*{../../CNV_analysis/Gste_diagnostic_read_CNVs.png}}
		\vskip 0.4cm
		\makebox{\includegraphics*{../../CNV_analysis/Cyp9k1_diagnostic_read_CNVs.png}}
	\end{center}
	\caption{\footnotesize Frequency (proportion of samples carrying at least one copy) of known CNV alleles detected using diagnostic reads around \textit{Ace1}, the \textit{Cyp6aa / Cyp6p} cluster, the \textit{Cyp6m / Cyp6z} cluster, the \textit{Gste} cluster and \textit{Cyp9k1}. Darkness of blue (\textit{An. gambiae}) and red (\textit{An. coluzzii}) provided as a visual aid for the magnitude of the value in each cell. The genomic coordinates of each CNV allele can be found in Supplementary Data @@. In each cluster, the ``Dup0'' column indicates the presence of increased copy number in any of the genes in the cluster. Where this is larger than the sum of known alleles, it suggests the presence of CNV alleles not present in the Ag1000G database. ``Del'' alleles in Ace1 represent secondary deletions within the Ace1-Dup1 CNV.}
	\label{FigS1}
\end{figure}

\clearpage

\begin{figure}[h]
	\includegraphics*[width = 7.9cm]{../../randomisations/Fst/Avrankou_coluzzii_Delta_peak_filter_plot.png}
	\vskip 0.4cm
	\includegraphics*[width = 7.9cm]{../../randomisations/Fst/Baguida_gambiae_Delta_peak_filter_plot.png}
	\includegraphics*[width = 7.9cm]{../../randomisations/Fst/Baguida_gambiae_PM_peak_filter_plot.png}
	\vskip 0.4cm
	\includegraphics*[width = 7.9cm]{../../randomisations/Fst/Korle-Bu_coluzzii_Delta_peak_filter_plot.png}
	\includegraphics*[width = 7.9cm]{../../randomisations/Fst/Korle-Bu_coluzzii_PM_peak_filter_plot.png}
	\vskip 0.4cm
	\includegraphics*[width = 7.9cm]{../../randomisations/Fst/Madina_gambiae_Delta_peak_filter_plot.png}
	\includegraphics*[width = 7.9cm]{../../randomisations/Fst/Madina_gambiae_PM_peak_filter_plot.png}
	\vskip 0.4cm
	\includegraphics*[width = 7.9cm]{../../randomisations/Fst/Obuasi_gambiae_Delta_peak_filter_plot.png}
	\includegraphics*[width = 7.9cm]{../../randomisations/Fst/Obuasi_gambiae_PM_peak_filter_plot.png}
	\caption{\footnotesize $F_{ST}$ between resistant and susceptible individuals in each sample set, calculated in 1000 SNP windows. Red line indicates $F_{ST}$, grey lines in background show results from 200 randomisations in which phenotype labels were permuted. Regions of extended haplotype homozygosity can cause spurious peaks in $F_{ST}$, which are captured by the randomisations (eg., peaks around Ace1 in Deltamethrin sample sets). Windows identified as peaks were considered ``significant'' (green points) if their $F_{ST}$ value fell above the 99th centile of the randomisations for that window (ie: P < 0.01) and non-significant (purple points) otherwise.}
	\label{FigS2}
\end{figure}

\clearpage

\begin{figure}[h]
	\includegraphics*[width = 7.9cm]{../../randomisations/H12/Avrankou.coluzzii.Delta_peak_filter_plot.png}
	\vskip 0.4cm
	\includegraphics*[width = 7.9cm]{../../randomisations/H12/Baguida.gambiae.Delta_peak_filter_plot.png}
	\includegraphics*[width = 7.9cm]{../../randomisations/H12/Baguida.gambiae.PM_peak_filter_plot.png}
	\vskip 0.4cm
	\includegraphics*[width = 7.9cm]{../../randomisations/H12/Korle-Bu.coluzzii.Delta_peak_filter_plot.png}
	\includegraphics*[width = 7.9cm]{../../randomisations/H12/Korle-Bu.coluzzii.PM_peak_filter_plot.png}
	\vskip 0.4cm
	\includegraphics*[width = 7.9cm]{../../randomisations/H12/Madina.gambiae.Delta_peak_filter_plot.png}
	\includegraphics*[width = 7.9cm]{../../randomisations/H12/Madina.gambiae.PM_peak_filter_plot.png}
	\vskip 0.4cm
	\includegraphics*[width = 7.9cm]{../../randomisations/H12/Obuasi.gambiae.Delta_peak_filter_plot.png}
	\includegraphics*[width = 7.9cm]{../../randomisations/H12/Obuasi.gambiae.PM_peak_filter_plot.png}
	\caption{\footnotesize Difference in H12 between resistant and susceptible sub-samples (resistant minus susceptible) in each sample set, calculated in 1000 SNP windows. Green line indicates H12.difference, grey lines in background show results from 200 randomisations in which phenotype labels were permuted. Regions of extended haplotype homozygosity can cause spurious peaks in H12.difference, which are captured by the randomisations. Windows identified as positive peaks were considered ``significant'' (green points) if their H12.difference value fell above the 99th centile of the randomisations for that window (ie: P < 0.01) and non-significant (purple points) otherwise.}
	\label{FigS3}
\end{figure}

\clearpage

\begin{figure}[h]
	\includegraphics*[width = 7.9cm]{../../randomisations/PBS/Avrankou.coluzzii.Delta_peak_filter_plot.png}
	\vskip 0.4cm
	\includegraphics*[width = 7.9cm]{../../randomisations/PBS/Baguida.gambiae.Delta_peak_filter_plot.png}
	\includegraphics*[width = 7.9cm]{../../randomisations/PBS/Baguida.gambiae.PM_peak_filter_plot.png}
	\vskip 0.4cm
	\includegraphics*[width = 7.9cm]{../../randomisations/PBS/Korle-Bu.coluzzii.Delta_peak_filter_plot.png}
	\includegraphics*[width = 7.9cm]{../../randomisations/PBS/Korle-Bu.coluzzii.PM_peak_filter_plot.png}
	\vskip 0.4cm
	\includegraphics*[width = 7.9cm]{../../randomisations/PBS/Madina.gambiae.Delta_peak_filter_plot.png}
	\includegraphics*[width = 7.9cm]{../../randomisations/PBS/Madina.gambiae.PM_peak_filter_plot.png}
	\vskip 0.4cm
	\includegraphics*[width = 7.9cm]{../../randomisations/PBS/Obuasi.gambiae.Delta_peak_filter_plot.png}
	\includegraphics*[width = 7.9cm]{../../randomisations/PBS/Obuasi.gambiae.PM_peak_filter_plot.png}
	\caption{\footnotesize PBS between resistant and susceptible individuals in each sample set, calculated in 1000 SNP windows. Orange line indicates PBS, grey lines in background show results from 200 randomisations in which phenotype labels were permuted. Regions of extended haplotype homozygosity can cause spurious peaks in PBS which are captured by the randomisations. Windows identified as positive peaks were considered ``significant'' (green points) if their PBS value fell above the 99th centile of the randomisations for that window (ie: P < 0.01) and non-significant (purple points) otherwise.}
	\label{FigS4}
\end{figure}

\clearpage

\begin{figure}[h]
	\hspace{-0.3cm}\includegraphics*[width = 8.4cm]{../supplementary_implicated_regions/Avrankou_coluzzii_Delta_implicated_regions.pdf}
	\vskip 0.4cm
	\hspace{-0.3cm}\includegraphics*[width = 8.4cm]{../supplementary_implicated_regions/Baguida_gambiae_Delta_implicated_regions.pdf}
	\hspace{-0.3cm}\includegraphics*[width = 8.4cm]{../supplementary_implicated_regions/Baguida_gambiae_PM_implicated_regions.pdf}
	\vskip 0.4cm
	\hspace{-0.3cm}\includegraphics*[width = 8.4cm]{../supplementary_implicated_regions/Korle-Bu_coluzzii_Delta_implicated_regions.pdf}
	\hspace{-0.3cm}\includegraphics*[width = 8.4cm]{../supplementary_implicated_regions/Korle-Bu_coluzzii_PM_implicated_regions.pdf}
	\vskip 0.4cm
	\hspace{-0.3cm}\includegraphics*[width = 8.4cm]{../supplementary_implicated_regions/Madina_gambiae_Delta_implicated_regions.pdf}
	\hspace{-0.3cm}\includegraphics*[width = 8.4cm]{../supplementary_implicated_regions/Madina_gambiae_PM_implicated_regions.pdf}
	\vskip 0.4cm
	\hspace{-0.3cm}\includegraphics*[width = 8.4cm]{../supplementary_implicated_regions/Obuasi_gambiae_Delta_implicated_regions.pdf}
	\hspace{-0.3cm}\includegraphics*[width = 8.4cm]{../supplementary_implicated_regions/Obuasi_gambiae_PM_implicated_regions.pdf}
	\caption{\footnotesize Genomic regions implicated in insecticide resistance by each of our four approaches. For the global GWAS method, these are 100,000 bp windows which contained at least 10 of the top 1000 significant SNPs. For $F_{ST}$, these are significant peaks which contained at least one haplotype significantly positively associated with resistance (Supplementary Data S2). For H12 and PBS, these are significant positive peaks (ie: indicating stronger signals of selection in resistant compared to susceptible samples).}
	\label{FigS5}
\end{figure}

\clearpage

\begin{figure}[h]
	\includegraphics*[width = 17cm]{../../CNV_analysis/Cyp9k1_Dup10_haplotypes/Cyp9k1_haplotype_clustering.png}
	\caption{\footnotesize Haplotypes bearing the CNV allele Cyp9k1\_Dup10 in \textit{An. gambiae} and \textit{An. coluzzii} form a single cluster nested within other \textit{An. coluzzii} haplotypes, indicating introgession of the allele from \textit{An. coluzzii} to \textit{An. gambiae}. Haplotype clustering was performed using 500 SNPs around \textit{Cyp9k1}. The Cyp9k1\_Dup10 CNV allele was phased by identifying SNPs perfectly correlated with its presence / absence level. Full workings to reproduce this analysis can be found at \url{https://github.com/vigg-lstm/GAARD\_work/blob/main/CNV\_analysis/Cyp9k1\_Dup10\_haplotypes/Cyp9k1\_CNV\_haplotypes.ipynb}.}
	\label{FigS6}
\end{figure}



\end{document}
